Include/object.h

```diff
@@ -303,7 +303,7 @@ _Py_ThreadId(void)
 static inline Py_ALWAYS_INLINE int
 _Py_IsOwnedByCurrentThread(PyObject *ob)
 {
-    return ob->ob_tid == _Py_ThreadId();
+    return _Py_atomic_load_uintptr_relaxed(&ob->ob_tid) == _Py_ThreadId();
```

Comment:
I think this is one of the few cases where we want to limit the use the atomic load to Tsan builds. Unfortunately, the atomic load generates code that's a bit worse on x86-64, and it can actually make a measurable difference because of how common reference counting is.

Something like:

```
#ifdef _Py_THREAD_SANITIZER
    return _Py_atomic_load_uintptr_relaxed(&ob->ob_tid) == _Py_ThreadId();
#else
    return ob->ob_tid == _Py_ThreadId();
#endif
```

Reply:
I was thinking about this a bit more, and I'm wondering how it's safe to use a non-atomic load here, at least in theory. Is there something that prevents the following from happening:

1. `ob->ob_tid` is compiled as two separate loads.
2. The execution of (1) races with a store of `0` to `ob->ob_tid` in another thread.
3. We end up with a torn read of `ob->ob_tid`, resulting in `ob->ob_tid == _Py_ThreadId()` incorrectly evaluating to true.

I know that this is probably exceedingly unlikely to happen in practice, but it seems like it's at least possible in theory.

Reply:
Yeah, it's a problem in theory, but I don't think it's a problem in practice (the loads will not be split) and the cost here is relatively high.

If we can get someone to make GCC and LLVM generate decent code on x86 with relaxed atomics, that would be ideal.

Another alternative is to do something like https://github.com/colesbury/nogil-3.12/blob/cedde4f5ec3759ad723c89d44738776f362df564/Include/pyatomic.h#L321-L337 and use inline asembly to generate the desired x86 code ourselves, but I think everyone would like to avoid inline assembly if possible.

Reply:
Makes sense!